### PR TITLE
8265206: Tree-/TableCell: editing state not updated on cell re-use

### PR DESCRIPTION
**Pull request by kleopatra**
Issue is missing update of Tree-/TableCell's editiable state when changing its index from editingIndex to -1. 

Seems to be a left-over from fixing cell's editing update - done in [JDK-8150525](https://bugs.openjdk.java.net/browse/JDK-8150525) - on index change for the special case of new index -1.

Fixed by cleaning out editing state in that corner case also, added tests that were failing before and passing after the fix. Note that there are also tests that passed before: the previous fix didn't add any tests, so added them here.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265206](https://bugs.openjdk.java.net/browse/JDK-8265206): Tree-/TableCell: editing state not updated on cell re-use

### Reviewers
 * [Ajit Ghaisas](https://openjdk.java.net/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Johan Vos](https://openjdk.java.net/census#jvos) (@johanvos - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/473/head:pull/473` \
`$ git checkout pull/473`

Update a local copy of the PR: \
`$ git checkout pull/473` \
`$ git pull https://git.openjdk.java.net/jfx pull/473/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 473`

View PR using the GUI difftool: \
`$ git pr show -t 473`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/473.diff">https://git.openjdk.java.net/jfx/pull/473.diff</a>

</details>

### CODE REVIEW

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableCell.java [hunk 1]
```diff
@@ -549,25 +549,40 @@ private void updateFocus() {
...
+     * editing state.
+     */
+    private void doCancelEdit() {
+        // If my index is not the one being edited then I need to cancel
+        // the edit. The tricky thing here is that as part of this call
+        // I cannot end up calling list.edit(-1) the way that the standard
+        // cancelEdit method would do. Yet, I need to call cancelEdit
+        // so that subclasses which override cancelEdit can execute. So,
+        // I have to use a kind of hacky flag workaround.
+        updateEditingIndex = false;
```
Comment (johanvos) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TableCell.java:581]: Do you need to assert on updateEditingIndex being true before this is done?

Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableCell.java:581]: hmm .. don't quite understand your comment: it's a flag to tell cancelEdit to not update control's editing state (that whole block is simply extracted from the pre-fix state .. where it has been for ages ;)

Comment (johanvos) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableCell.java:581]: I see the updateEditingIndex is used as a hacky flag as the comment says, and `table.edit(-1, null)` is called conditionally.
But my question is about the pre- and postconditions. What would happen if a subclass overrides `cancelEdit` and throws an exception, so that `updateEditingIndex = true` is not called after the `cancelEdit`? In that case, `updateEditingIndex` stays false.
I'm not saying this is an issue, just wondering about this as this flag does an important thing -- and I do know this was in the code before the PR, so the PR is definitely an improvement, I'm just thinking that this might be an opportunity to deal with the precondition that updateEditingIndex *should* be true when entering the method and false when leaving the method.

Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableCell.java:581]: thanks for the explanation and spelling it out to me :)

> But my question is about the pre- and postconditions. What would happen if a subclass overrides `cancelEdit` and throws an exception, so that `updateEditingIndex = true` is not called after the `cancelEdit`? In that case, `updateEditingIndex` stays false.

talking about pre/postCondiditions: wouldn't an override of cancelEdit (and the other editing life-cycle methods) that's throwing an exception violate its contract? Don't see any constraints in its spec at Cell, so I would say it must do it's very best not to throw anything, at least not intentionally. And if it happens somewhere along the chain, f.i.  an edit handler going wild, then it would be the responsibility of that handler to behave (aka: usage error except for the little bugs in the editEvents .. ). 

> I'm not saying this is an issue, just wondering about this as this flag does an important thing -- and I do know this was in the code before the PR, so the PR is definitely an improvement, I'm just thinking that this might be an opportunity to deal with the precondition that updateEditingIndex _should_ be true when entering the method and false when leaving the method.

hmm .. again not quite sure I understand: shouldn't updateEditingIndex be true both on entering and leaving doCancelEdit?. Anyway, if there's something going wrong in cancelEdit, the cell is in an arbitrary state: could still be editing (if the wrong-going happened before calling cell.cancelEdit) or not, could have fired the cancel event or not, could have updated its visuals or not .. there's not much we can do to recover from it. When going from this undeterminate state into the next editing cycle the outcome is .. unpredictable. 

All that said: we could wrap the flag un-/setting into a try-finally block - then at least the whacky flag is in a specific state after all hell broke loose ..

Comment (johanvos) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableCell.java:581]: I agree a try-finally block here (setting the flag in a specific state) might be useful -- I prefer to keep unpredictable states for quantum computing :)

Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableCell.java]: done for Tree/TableCell: 

- wrapped into try-finally block with resetting updateEditingIndex flag to true in finally
- added test to Tree/TableCellTest that fails before and passes after the change

now the flag is more predictable than the behavior of extending classes' cancelEdit implemenations (and states in quantum computing - probably, your book is still on my reading list :)

**Review by aghaisas — APPROVED** [review]

**Review by johanvos — APPROVED** [review]
